Tutorial: Using HogQL for advanced time and date filters (#6006)
PostHog/posthog.com: contents/tutorials/hogql-date-time-filters.md

---
title: Using HogQL for advanced time and date filters
date: 2023-05-19
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/tutorial-9.png
tags: ["hogql", "insights"]
---

> Because there are infinite ways to break down time, there are infinite ways to filter based on time. HogQL unlocks more of these in PostHog, and in this tutorial, we go through examples of how to use do that.

Reviewer: ```suggestion
Because there are infinite ways to break down time, there are infinite ways to filter based on time. HogQL unlocks more of these in PostHog, and in this tutorial we'll go through examples of how to use do that.
```

To add a HogQL filter, create a new insight, open the filter dropdown, click "Add filter" below your data series, select HogQL from the options, write your expression, and click "Add HogQL expression." HogQL filters are available on every type of insight from trends to funnels to lifecycle.

![hogql.mp4](../images/tutorials/hogql-date-time-filters/hogql.mp4)

## Accessing your data’s dates and times

> To filter on dates and times, you need access to those values. You can find a full list of events, properties, and types in your [data management tab](https://app.posthog.com/data-management/events), but here is a summary of the important points:

Reviewer: I dont think we need to say "To filter on dates and times, you need access to those values."

```suggestion
Below is frequently used time properties (You can find a full list of events, properties, and types in your [data management tab](https://app.posthog.com/data-management/events)):
```

> - You can access the time events happened with the `timestamp` property which is automatically set on events.

Reviewer: ```suggestion
- The `timestamp` property indiciates what time an event occurerd, and is automatically set.
```

> - Persons have a `created_at` property which you can access with `person.created_at`.

Reviewer: When referring to posthog concept, like persons, for the first time, we should hyper link to it

```suggestion
- [Persons](/docs/data/persons) have a `created_at` property which you can access with `person.created_at`.
```

> - For custom event or person properties, such as `subscribed_at`, you can access them with `person.properties.subscribed_at`.

Reviewer: ```suggestion
- For custom events or person properties, you can access them with `person.properties.{custom_property_name}`.
```

> **Note:** you might need to convert strings to date time objects with the `toDateTime()` function or convert date time objects to strings with `toString()`.

> The [ClickHouse SQL statements](https://clickhouse.com/docs/en/sql-reference) we built HogQL on also have useful helper functions to know when it comes to using dates. These include:

Reviewer: ```suggestion
The [ClickHouse SQL statements](https://clickhouse.com/docs/en/sql-reference) we built HogQL on also have useful helper functions that are good to know when working with dates. These include:
```

- `now()`: the current date and time at the moment of query analysis.
- `today()`: the current date at the moment of query analysis.

> - `yesterday()`: yesterday’s date at the moment of query analysis.

Reviewer: should we include a link to a more comprehensive list of helper functions related to date+time?

Author: The problem is that:
1. There is a lot with varying degrees of relevance.
2. Not all of the ClickHouse SQL functions/helpers are implemented in HogQL.

## Events from a specific time range

To filter for events in a specific time range, you can use `toDateTime()` to create a date time object that we compare with the event `timestamp`. 

> For example, if we wanted to filter for events after 9:26 AM on September 16th, 2022, and before 2:34 PM on October 1st, 2022, we can use the expression:

Reviewer: Nit: we're mostly using `you` instead of `we` throughout, so we should stick to that.

Nit:

```suggestion
For example, if you want to filter for events `after 9:26 AM on September 16th, 2022, and before 2:34 PM on October 1st, 2022`, we can use the expression:
```

```
toDateTime('2022-09-16 09:26:00') < timestamp 
and timestamp < toDateTime('2022-10-01 14:34:00')
```

> **Note:** by default, PostHog uses `UTC` for our timestamps, so you should keep that in mind and convert times to `UTC` where relevant.

## Using the Unix timestamp

> If you have a Unix timestamp to filter by, you can use `toUnixTimestamp` to convert a date time object into a Unix timestamp, or `fromUnixTimestamp` to do the reverse. 

Reviewer: ```suggestion
If you'd like to filter using a Unix timestamp, you can use `toUnixTimestamp` to convert a date time object into a Unix timestamp, or `fromUnixTimestamp` to do the reverse. 
```

For example, if you wanted to filter for events after the Unix timestamp `1674259200`, you could use the expression:

```
fromUnixTimestamp(1674259200) < timestamp
```

## Older or younger events

If you are looking for events of a certain age, rather than a relation to a certain date, you can use the `dateDiff()` function with the timestamp and current time. 

For example, to get events over 5 days old, use the expression:

```
dateDiff('day', timestamp, now()) > 5
```

You can replace `day` here with `second`, `minute`, `hour`, `week`, `month`, `quarter`, and `year` to get the differences in those values as well.

## Weekly and quarterly reports

> Many companies report on a weekly or quarterly basis. For this, standard 7 or 90-day date range filters may include values beyond the days or weeks you want to report. HogQL lets us easily create these reports.

Reviewer: > For this, standard 7 or 90-day date range filters may include values beyond the days or weeks you want to report.

This sentence is not clear. I'm not sure what it means.

Author: Rewrote

For example, if we wanted events for this quarter, we can use the expression:

```
toStartOfQuarter(timestamp) == toStartOfQuarter(now())
```

We can do the same with the current week with `toStartOfWeek`:

```
toStartOfWeek(timestamp) == toStartOfWeek(now())
```

Alternatively, if you wanted a weekly report for only weekdays, you can use `toDayOfWeek` to filter out Saturday and Sunday with their day’s number:

```
toDayOfWeek(timestamp) != 6 and toDayOfWeek(timestamp) != 7

> ```

Reviewer: Let's include a link somewhere here to the clickhouse docs for these functions e.g., https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#tostartofquarter

## Analyzing subscribers or trial users

> HogQL lets you analyze users based on dates too. This enables you to filter for events based on users in trial or recently subscribed.

Reviewer: > HogQL lets you analyze users based on dates too.

What do you mean by this? Isn't this what we've been doing in the entire tutorial?

Author: rewrote

For example, you want to events from users in the last 3 days of their trial period and you only have a `trial_started` person property. You can use the `addDays()` function to add 30 days and check if that date is less than or equal to 3 days away from `now()` like this: 

```
dateDiff(
	'day', 
	now(), 
	addDays(toDateTime(person.properties.trial_started), 30)

> ) <= 3

Reviewer 2: The `addDays` part is often simpler with `interval` like `toDateTime(person.properties.trial_started) + interval 30 day`.

Also, if the person property is set as a `DateTime` type from the "data management" pages, you don't need to wrap it with `toDateTime`.

---
title: Using HogQL for advanced time and date filters
date: 2023-05-24
author: ["ian-vanagas"]
showTitle: true
sidebar: Docs
featuredImage: ../images/tutorials/banners/tutorial-9.png
tags: ["hogql", "insights"]
---

> Because there are infinite ways to break down time, there are infinite ways to filter based on time. HogQL unlocks more of these in PostHog, and in this tutorial we'll go through examples of how to use do that.

Reviewer: nit: feel free to ignore
```suggestion
Since there are infinite ways to break down time, there are infinite ways to filter based on time. HogQL unlocks more of these in PostHog, and in this tutorial we'll go through examples of how to use do that.
```

To add a HogQL filter:

> 1. Create a new insight

Reviewer: nit: hyperlink

```
toDateTime('2022-09-16 09:26:00') < timestamp 
and timestamp < toDateTime('2022-10-01 14:34:00')
```

> **Note:** by default, PostHog uses `UTC` for timestamps.

## Using the Unix timestamp

If you'd like to filter using a Unix timestamp, you can use []`toUnixTimestamp`](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#tounixtimestamp) to convert a date time object into a Unix timestamp, or `fromUnixTimestamp` to do the reverse. 

For example, if you wanted to filter for events after the Unix timestamp `1674259200`, you could use the expression:

```
fromUnixTimestamp(1674259200) < timestamp
```

## Relative time filters

If you'd like to query events relative to today's date (or any other date), you can use the [`dateDiff()`](https://clickhouse.com/docs/en/sql-reference/functions/date-time-functions#date_diff) function. 

For example, to get events older than 5 days, use the expression:

```
dateDiff('day', timestamp, now()) > 5
```

You can also replace `day` with `second`, `minute`, `hour`, `week`, `month`, `quarter`, and `year`.

## Weekly and quarterly reports

> Many companies report on a weekly or quarterly basis. Filtering for events in the last 7 or 90 days may include ones outside of the current week or quarter, making these reports inaccurate. HogQL enables us easily create reports for specific weeks and quarters.

Reviewer: ```suggestion
Many companies report on a weekly or quarterly basis. Using a simple filter, such `events in the last 7 days` or `events in the last 90 days` will likely include events from the previous week or quarter, making these reports inaccurate. Thus HogQL provides filters for easily querying events from specific weeks and quarters.
```
